packages/core/saved-objects/core-saved-objects-migration-server-internal/src/document_migrator/pipelines/upgrade_pipeline.ts

```diff
@@ -7,38 +7,59 @@
  */
 
 import Boom from '@hapi/boom';
+import { cloneDeep } from 'lodash';
 import Semver from 'semver';
 import type { SavedObjectUnsanitizedDoc } from '@kbn/core-saved-objects-server';
-import { ActiveMigrations, Transform, TransformType } from './types';
-import { maxVersion } from './utils';
+import { ActiveMigrations, Transform, TransformType } from '../types';
+import type { MigrationPipeline, MigrationPipelineResult } from './types';
+import {
+  coreVersionTransformTypes,
+  applyVersion,
+  assertValidCoreVersion,
+  maxVersion,
+} from './utils';
 
 function isGreater(a?: string, b?: string) {
   return !!a && (!b || Semver.gt(a, b));
 }
 
-export class DocumentMigratorPipeline {
-  additionalDocs = [] as SavedObjectUnsanitizedDoc[];
-
-  constructor(
-    public document: SavedObjectUnsanitizedDoc,
-    private migrations: ActiveMigrations,
-    private kibanaVersion: string,
-    private convertNamespaceTypes: boolean
-  ) {}
+export class DocumentUpgradePipeline implements MigrationPipeline {
+  private additionalDocs = [] as SavedObjectUnsanitizedDoc[];
+  private document: SavedObjectUnsanitizedDoc;
+  private originalDoc: SavedObjectUnsanitizedDoc;
+  private migrations: ActiveMigrations;
+  private kibanaVersion: string;
+  private convertNamespaceTypes: boolean;
+  private targetTypeVersion: string;
+
+  constructor({
+    document,
+    migrations,
+    kibanaVersion,
+    convertNamespaceTypes,
+    targetTypeVersion,
+  }: {
+    document: SavedObjectUnsanitizedDoc;
+    migrations: ActiveMigrations;
+    kibanaVersion: string;
+    convertNamespaceTypes: boolean;
+    targetTypeVersion?: string;
```

Comment:
Added the `targetTypeVersion` option to the up pipeline, as I suspect we will need it. Note that atm, the option is not exposed or used by the document migrator itself.